pandas/core/arrays/base.py

```diff
@@ -236,6 +237,52 @@ def isna(self):
         """
         raise AbstractMethodError(self)
 
+    def _values_for_argsort(self):
+        # type: () -> ndarray
+        """Get the ndarray to be passed to np.argsort.
```

Comment:
`_ndarray_valaues` seems like more of an internal thing, no? I don't know enough to say whether the `_ndarray_values` appropriate for our current uses (mostly indexing IIRC) are also appropriate for argsort.

Reply:
my point is what is the point of overriding this specific one? why is not a general purpose EA method/property used here. The proliferation of methods properties is really troublesome.

Reply:
if someone wants to override argsort great. but providing an indirect mechism is really kludgey.

Reply:
> my point is what is the point of overriding this specific one?

This PR is implementing argsort. I could see a similar pattern for factorize.

> The proliferation of methods properties is really troublesome.

How so?

> but providing an indirect mechism is really kludgey.

What's kudgey about it? I think the common case will be overriding the values provided to `np.argsort`, not overriding the sorting algorithm itself. This is true for Categorical and IPArray, and will be true for Period and probably others.

Without `_values_for_argsort` we, and 3rd party libraries, will have duplicate code for validating keyword arguments and the ascending kwarg.

Reply:
why is this different from ``_ndarray_values``, which is implemented on EA arrays? having 2 is just plain confusing, not to mention you have a ``_formatting_values```.

Reply:
FYI, `_ndarray_values` is present on *containers* like Series / Index. We don't have a similar convention for the arrays themselves.

`_formatting_values` is specific for printing, which we certainly don't want here.

I'm going to do something like this when moving Period to an EA, otherwise we'll have duplicate logic for parameter validation and the ascending in 3 places. It's really just a matter of whether we want to expose it to library authors implementing an EA.

Reply:
my point is that I don't want a specific ``_values_for_argsort`` but rather something *like* ``_ndarray_values``, and I would just call it that. you need to turn this into an ndarray sometimes, so provide a methdo/property for that.
OR simply call ``np.asarray()`` if you actually need an ndarray.

Reply:
We don't necessarily want to convert into extension arrays into the same NumPy array used for `np.asarray()`.

For example, the IP Address extension array probably wants to convert into numpy object array of IPAddress object. But for sorting, it could just return numpy structured array with a few integer fields, which will be much faster for comparisons than Python IPAddress objects.

Reply:
Indeed. In the IPArray case, there's a numpy array, so `_values_for_argsort` is just that array.

>  OR simply call np.asarray() if you actually need an ndarray.

Then we're back in the duplicate code situation. That's OK for the base class, but Categorical, Period, and Interval will end up re-implementing argsort from scratch. With `_values_for_argsort`, it's just a matter of constructing that array (codes, ordinals, concat left & right).

Reply:
On the name `_values_for_argsort`, it's *possible* that we'll find other uses for it, in which case we just change the name.

Do you know if a simple array appropriate for arg-sorting is also appropriate for factorize, joins, indexing, etc? I'm not sure ahead of time.

Reply:
> Do you know if a simple array appropriate for arg-sorting is also appropriate for factorize, joins, indexing, etc?

Well, we can scratch factorize / groupby off. Categorical defines `_codes_for_groupby` separately from its `ndarray_values` (codes) https://github.com/pandas-dev/pandas/blob/3783ccc694ef93c0cf159ca1ad8e69a67d828bf0/pandas/core/arrays/categorical.py#L638

So we can't just use one array for everything.

Reply:
ok, so then let's pick a name change ``_codes_for_groupby``. in this refactor we want to find other usecases and fix our code now rather than later.

something like: ``_int_mapping_for_values``